modules/calendar/add_edit_event.php

```diff
@@ -1337,20 +1337,36 @@ function find_available(extra) {
       /***************************************************************
       $qsql = sqlStatement("SELECT * FROM facility WHERE service_location != 0");
       ***************************************************************/
-      $facils = getUserFacilities($_SESSION['authId']);
+      $facils = getUserFacilities($userid);
+      //userid is id of clicked user(provider) in calendar
+      //ufid is an array of id of schedule facilities of selected user
+      $ufid = array();
+      foreach ($facils as $uf) {
+        $ufid[] = $uf['id'];
+      }
       $qsql = sqlStatement("SELECT id, name FROM facility WHERE service_location != 0");
       /**************************************************************/
       while ($facrow = sqlFetchArray($qsql)) {
         /*************************************************************
         $selected = ( $facrow['id'] == $e2f ) ? 'selected="selected"' : '' ;
         echo "<option value={$facrow['id']} $selected>{$facrow['name']}</option>";
         *************************************************************/
-        if ($_SESSION['authorizedUser'] || in_array($facrow, $facils)) {
+        if ($GLOBALS['restrict_user_facility']) {
+          //if restricting users(providers) to schedule facilities
+          //then list only schedule facilities of that user or default facility
+          //in Facility in calendar's event panel
+          //e2f denotes id of default facility
+          if (in_array($facrow['id'], $ufid) || $facrow['id'] == $e2f || $facrow['id'] == 3) {
```

Comment:
id 3 is the default facility, yes, which is generally edited, but the "default facility" should be the *users* default, right?  Generally `facilities.id` == 3 is a facility that gets edited (name changed), but there is no guarantee that a user will have access to it, or that it will be the primary business entity, or any of that.

-example use of these restrictions, and thus the main goal here:
1.  There is a central office.  Folks here generally have access to everything.
2.  There are "nursing homes" or "assisted living facilities" or other locations that either have provider staff (nurses, home care person, etc...) that should only every be able to see the patients at *that* facility.  There may be a hundred or more of these satellite facilities.
3.  Some of the providers see patients at more than one facility.  There may not be any particular apparent rhyme or reason to the combinations.

Use of this in combination with TAGS:
1.   in patient_data, a new field 'facility` is added.  This is a dropdown listing all the facilities.  It is (should be) set as a *required* field for this purpose (configuring required fields is another area of concerns).  This is a drop-down list of facilities, which should only list facilities that the USER has access to.
2.  A TAG is created, wherein a user has restricted access, then allowed access based the value in `patient_data.facility` or on a patient tag that does pretty much the same thing.   This actually becomes a set of tags that deny, then allow.  Insert examples follow:
```
Facility ID Tag, Deny Tag, Allow Tag

INSERT INTO `tf_patients_tags` (`id`, `tag_id`, `pid`, `created_at`, `created_by`, `updated_at`, `updated_by`, `status`) VALUES
(12443, 70, 207578, '0000-00-00 00:00:00', '0000-00-00 00:00:00', '0000-00-00 00:00:00', '0000-00-00 00:00:00', '0'),
(24876, 1, 207578, '0000-00-00 00:00:00', '0000-00-00 00:00:00', '0000-00-00 00:00:00', '0000-00-00 00:00:00', '0'),
(37309, 2, 207578, '2018-04-26 11:05:21', '0000-00-00 00:00:00', '2018-04-26 11:05:21', '0000-00-00 00:00:00', '1');

Deny Tag, Allow Tag

INSERT INTO `tf_filters` (`id`, `created_at`, `created_by`, `updated_at`, `updated_by`, `requesting_action`, `requesting_type`, `requesting_entity`, `object_type`, `object_entity`, `note`, `gacl_aro`, `gacl_acl`, `effective_datetime`, `expiration_datetime`, `priority`, `deleted`) VALUES
(10609, '0000-00-00 00:00:00', 'admin', '0000-00-00 00:00:00', 'admin', 'deny', 'user', 'admin', 'tag', 1, '', 0, 0, '0000-00-00 00:00:00', '0000-00-00 00:00:00', 3, 0),
(10610, '2018-05-22 00:00:00', 'admin', '2018-04-26 00:00:00', 'admin', 'allow', 'user', 'admin', 'tag', 2, '', 0, 0, '0000-00-00 00:00:00', '0000-00-00 00:00:00', 1, 0);

Facility Tag, Deny Tag

INSERT INTO `tf_filters` (`id`, `created_at`, `created_by`, `updated_at`, `updated_by`, `requesting_action`, `requesting_type`, `requesting_entity`, `object_type`, `object_entity`, `note`, `gacl_aro`, `gacl_acl`, `effective_datetime`, `expiration_datetime`, `priority`, `deleted`) VALUES
(6476, '0000-00-00 00:00:00', 'admin', '0000-00-00 00:00:00', 'admin', 'allow', 'user', 'aartis', 'tag', 119, '', 0, 0, '0000-00-00 00:00:00', '0000-00-00 00:00:00', 1, 0),
(10185, '0000-00-00 00:00:00', 'admin', '0000-00-00 00:00:00', 'admin', 'deny', 'user', 'aartis', 'tag', 1, '', 0, 0, '0000-00-00 00:00:00', '0000-00-00 00:00:00', 3, 0);

```
In the calendar, all we are worried about is the schedule_facilities...I think.  Essentially, by doing it this way we are using the following hacks:
1.  Using a relatively complex set of TAGS that users will need to replicate.
2.  Using the schedule_facilities feature, which is supposed to restrict what a user can *schedule for* (different than "see") as an access control feature. 

The original implementation of this was the Facility ACL, that basically used the patient's custom `facility` field as the check against the user's schedule facilities to determine access.  This system was a "plugin" by Ken Chapple.  Unfortunately, the custom facility_acl plugin, and the merged "tags" plugin used functions with identical names, making it a mod, and not a plugin.  The hooks were in the right place and all, but there was considerable modification to get the two to co-exist.  The attempt to get the code reviewed and merged fail...two or three times.  By the third attempt, too many things had changed, and I could never get the third version to work along side the tags, due to incompetence, spaghetti code, and lack of will.

Doing this in this manner is fine...for now, but we should do a total ACL replacement that includes per-patient-per-facility-per-user capabilities, along with "grouping" based on very clean and intuitive configuration UI components.  My guess is that it should be a native system, though I will admit that GACL was probably capable of doing all this, and the implementation was just half-baked.